Add utility class `Nuclide` to build nuclide strings, and return Z and A values

lemointm opened the pull request:
As a helpful utility, the `radioactivity.Nuclide` class is added to build an object from atomic number and atomic mass, or element/nuclide name. If atomic mass is omitted, the function returns the element symbol corresponding to the input atomic number. Also included are properties to return Z and A values for the element/nuclide (A = 0 is returned for elements).

lemointm commented:
Please disregard this PR, as changing the `DecayData` class to hold a `DataFrame` will resolve this issue.

lemointm commented:
Reopening this to add simple Z, A parsing without the need for major refactoring.

alexmalins commented:
Hi @lemointm 

There are some really great ideas in this PR! I think it would be great if we could do the following:

- Since you added the stable nuclides, I think the Radionuclide class should be subsumed into this new Nuclide class. All the functionality of Radionuclide works for the stable nuclides, so renaming to Nuclide will make it clearer that stable nuclides are supported.
- The new Z and A properties are great. What about adding an atomic_mass property or method as well? To return the atomic mass in g/mol?
- As mentioned on Slack, reading the csv file on every instantiation is inefficient. Actually, given Wolfgang's suggestions and the fact that the package file I/O's might have to be overhauled in future, I was thinking instead of reading a CSV of the element symbol data, it might just be easier to just write this into a list or dict into the code itself. It's not so big a list to do this.
- In terms of functionality, the class you drafted doesn't appear to support metastable state specification. I was thinking for increasing the general usability of the package, it would be good if support was added for PyNE-style `zzzaaassss` canonical [ids](https://pyne.io/theorymanual/nucname.html). This also has the advantages that the __init__() arguments for Nuclide and Inventory will become simpler - no need to have the second and third int parameters for A and state number that are only sometimes used. We'd then need some function or method to convert nuclide name strings to zzzaaassss, On the TARDIS side zzzaaassss ints will be easy to process to get what you need just with `//` integer divisions.

I'm happy to guide you in coding this, and writing the docs and tests!

lemointm commented:
I added most of the functionality, although it was hasty and may have some bugs; next I'll get on to doc updates and tests! Let me know if you find anything that could be improved.

lemointm commented:
Oh one tiny thing - I noticed when I added the `Radionuclide` methods to `Nuclide` that the plot method is again adding extra white space, I may have not updated my code in this branch to include the edits you made to fix this, I can double check.

alexmalins commented on radioactivedecay/__init__.py:
```diff
@@ -15,3 +15,4 @@
 from radioactivedecay.decaydata import DecayData, DEFAULTDATA
 from radioactivedecay.radionuclide import Radionuclide
```
We'll need to delete this line and remove/replace all references to Radionuclide in the code and tests to Nuclide.

alexmalins commented on radioactivedecay/decaydata.py:
```diff
@@ -242,6 +242,8 @@ class DecayData:
         NumPy array of dictionaries with direct progeny as keys and lists with branching fraction
         and decay mode data as values.
+    masses : numpy_ndarray
+        NumPy array of molar masses of the nuclides in the dataset.
```
I think this will disappear when you rebase to the latest devel branch.

alexmalins commented on radioactivedecay/element_symbols.csv:
```diff
@@ -0,0 +1,119 @@
+Z,Symbol
```
This file can be deleted now the data is going to be written into the source.

alexmalins commented on radioactivedecay/utils.py:
```diff
@@ -17,7 +17,7 @@
 def parse_nuclide(nuclide: str) -> str:
     """
-    Parses a nuclide string from e.g. '241Pu' or 'Pu241' format to 'Pu-241' format. Not this
+    Parses a nuclide string from e.g. '241Pu' or 'Pu241' format to 'Pu-241' format. Note this
```
Thanks!

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+    _check_fig_ax,
+)
+from radioactivedecay.utils import parse_radionuclide
+from radioactivedecay.utils import parse_nuclide
```
You can merge these two import statements into one line using a comma.

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+        For atomic number as input, atomic mass can be optionally
+        added for full nuclide string construction. 
+
+    Properties
```
These should be referred to as Attributes rather than Properties - https://numpydoc.readthedocs.io/en/latest/format.html#documenting-classes

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+            112: "Cn", 113: "Nh", 114: "Fl", 115: "Mc", 116: "Lv", 117: "Ts",
+            118: "Og"
+        }
+        sym_dict = dict((v,k) for k,v in Z_dict.items())
```
For Z_dict and sym_dict, as they don't ever change, it would be more efficient to create them once rather than every time a Nuclide is instantiated. I think a good place to put them would be in either the utils or the decaydata modules, where similar information is stored. You can put them in as constants, and import them at the top of this module.

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+        ) -> None:
+        
+        if A == 0:
+            if type(vary_input) == int:
```
isinstance() is a more accepted way of doing this than type() ==

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+    """
+    
+    def __init__(
+        self, input: Any,
```
Union[str, int] is more descriptive than Any here

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+        A: int = 0,
+        meta_state: str = "",
+        data: DecayData = DEFAULTDATA
+        ) -> None:
```
Ok let's discuss this one on Zoom. I've an uneasy feeling about how the constructor takes basically takes one argument is supplying a nuclide string, whereas three arguments for the Z, A, meta_state input. This has an un-pythonic tint, which is why I suggested just have a single `input`/`nuclide` argument that could be a nuclide string or zzzaaammmm int as a solution. Is there a specific need/reason you had why including the Z, A, meta_state input format is needed?

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+    
+    def __repr__(self) -> str:
+        if self.A == 0:
+            rep =  "Element: " + self.nuclide_name
```
Likewise this felt a little un-pythonic, how the Nuclide class can also behave like an Element class. A more natural way would be to have a separate Element class if needed. But do we need an Element class? Did you have a use case in mind when you made this?

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+    name : str
+        Nuclide string if A data is given, otherwise the element
+        symbol is returned.
+    Z : int
```
Z -> Z_val
(Attributes should be a list of all the data held in self. variables)
& A -> A_val below

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+        """
+        
+        if self.nuclide_name not in self.data.radionuclides:
+            raise ValueError(
```
I think this shouldn't need to be checked here, as we should be checking the nuclide exists in the dataset upon instantiation of the object.

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+            )
+        
+        index = self.data.radionuclide_dict[self.nuclide_name]
+        mass = self.data.masses[index]
```
Apologies, this will have to be updated due to refactors in the decaydata storage code. The data is now held in data.scipy_data.atomic_masses

alexmalins commented on radioactivedecay/nuclide.py:
```diff
@@ -0,0 +1,651 @@
+        self.Z_dict = Z_dict
+        self.sym_dict = sym_dict
+        self.parse_name(input, A, meta_state)
+        self.data = data
```
Just FYI I renamed data -> decay_data when I refactored a load of stuff recently. Was trying to improve areas where my code was previously weaker, i.e. descriptiveness of variable names

alexmalins reviewed (commented):
Ok Thom, good progress. As you've realized, a rebase is now in order to bring everything in line with the recent updates to devel. Note I changed a few variable names to make them more descriptive/intuitive, so this might mean you have a bit of tedious work updating all the stuff you copied over from the old version of radionuclide.py - apologies for this.

Moving forward, what we need next is to remove all the deprecated radionuclide/Radionuclide references and update to Nuclide. This will require:
- Renaming test_radionuclide.py to test_nuclide.py. Putting new tests in for your new functionality (but let's finalize the new API first to save duplicated effort before you write the new tests).
- Renaming all imports and references to `Radionuclide` to `Nuclide` in the other source and test files.
- Updating the docs and the readme, switching Radionuclide to Nuclide.
The search function in VSCode will be really helpful for making these changes.

But, yep, let's speak about the API for Nuclide instantiation first as this is the main comment I have. In fact, we could naturally finish and merge this PR just with nuclide string as accepted parameters, and then do the zzzaaammmm stuff in a seprate PR. That would make a lot of sense actually.

lemointm commented:
There are still some a few changes I need to make, namely merging devel to get your changes to Radionuclide that will be put into nuclide, and fixing tests, but the latest commit has some nice fixes to get around the element problem (namely, element symbol parsing is put in utils so that when it's needed for TARDIS we can use the utils and not have element parsing needed in Nuclide). Let me know what you think!

lemointm commented:
Here's the merged changes, this adds your devel changes to nuclide at least... next is fixing tests and docs!

alexmalins commented:
Here's some small comments I noticed while looking through:

- decaydata.py has picked up an old part of the docstring on lines 344-345 - the atomic masses are now in the DecayMatrices class - so those lines can be deleted
- nuclide.py in the attributes section of the Nuclide docstring, list all the attributes (add the atomic_mass, prog_bf_modes and decaydata attributes)
- L67 __init__() now raises TypeError instead of ValueError
- In the refactoring, the __eq__, __ne__ and __hash__ methods have been deleted - did you intend this?
- util.py L142 - would be good to have a docstring for build_nuclide_string
- Radionuclide -> Nuclide on L160

Going well so far! Thanks

lemointm commented:
I think it should all be working properly now, and I fixed up the loose ends with Radionuclide references and such. Docs are next!

alexmalins reviewed (commented):
Going great, thanks again. These are suggestions only - it's no problem if you're busy with school and don't have time to do them:

- It would be useful if the Inventory's supported the zzzaaassss canonical ids as well. Getting this working will require updates around instantiation and the `add()`, `subtract()` and `remove()` methods. So not a simple one liner. Ideally it would reuse the new `Nuclide.parse_name()` code you wrote. Obvious place to store the code applicable for both classes would be in utils.py, then have `Nuclide` and `Inventory` call that function.
- I see you've got all the existing tests up and running, but have you thought of adding some new tests to protect and verify your new code in nuclide.py and utils.py?

lemointm commented:
I added the id parser to utils, and actually placed it in inside `parse_nuclide` as well so that most of the references in `Inventory.py` were already good to go! I also added a good amount of tests for both `utils.py` and `Nuclide.py`, let me know if I missed any functionality.

alexmalins reviewed (approved):
Superb work - thank you @lemointm
